module/zfs/zfs_ioctl.c

```diff
@@ -5232,25 +5499,992 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
-	ASSERT3U(ioc, <, ZFS_IOC_LAST);
-	ASSERT3P(vec->zvec_legacy_func, ==, NULL);
-	ASSERT3P(vec->zvec_func, ==, NULL);
+	error = nvlist_lookup_int32(innvl, "fd", &fd);
+	if (error != 0)
+		return (SET_ERROR(EINVAL));
 
-	vec->zvec_legacy_func = func;
-	vec->zvec_secpolicy = secpolicy;
-	vec->zvec_namecheck = namecheck;
-	vec->zvec_allow_log = log_history;
+	error = dsl_pool_hold(snapname, FTAG, &dp);
+	if (error != 0)
+		return (error);
+
+	error = dsl_dataset_hold(dp, snapname, FTAG, &ds);
+	if (error != 0) {
+		dsl_pool_rele(dp, FTAG);
+		return (error);
+	}
+
+	mutex_enter(&ds->ds_sendstream_lock);
+
+	/*
+	 * Iterate over all the send streams currently active on this dataset.
+	 * If there's one which matches the specified file descriptor _and_ the
+	 * stream was started by the current process, return the progress of
+	 * that stream.
+	 */
+	for (dsp = list_head(&ds->ds_sendstreams); dsp != NULL;
+	    dsp = list_next(&ds->ds_sendstreams, dsp)) {
+		if (dsp->dsa_outfd == fd &&
+		    dsp->dsa_proc->group_leader == curproc->group_leader)
+			break;
+	}
+
+	if (dsp != NULL)
+		fnvlist_add_uint64(outnvl, "offset", *(dsp->dsa_off));
+	else
+		error = SET_ERROR(ENOENT);
+
+	mutex_exit(&ds->ds_sendstream_lock);
+	dsl_dataset_rele(ds, FTAG);
+	dsl_pool_rele(dp, FTAG);
+	return (error);
+}
+
+static int
+zfs_stable_ioc_promote(const char *fsname, nvlist_t *innvl, nvlist_t *outnvl,
+    nvlist_t *opts, uint64_t version)
+{
+	char parentname[MAXNAMELEN];
+	char conflsnap[MAXNAMELEN];
+	dsl_pool_t *dp;
+	dsl_dataset_t *clone;
+	dsl_dataset_t *origin = NULL;
+	dsl_dir_t *dd;
+	int error;
+
+	error = dsl_pool_hold(fsname, FTAG, &dp);
+	if (error != 0)
+		return (error);
+
+	error = dsl_dataset_hold(dp, fsname, FTAG, &clone);
+	if (error == 0) {
+		dd = clone->ds_dir;
+		error = dsl_dataset_hold_obj(dd->dd_pool,
+		    dd->dd_phys->dd_origin_obj, FTAG, &origin);
+		if (error != 0) {
+			dsl_dataset_rele(clone, FTAG);
+			dsl_pool_rele(dp, FTAG);
+			return (error);
+		}
+
+		dsl_dataset_name(origin, parentname);
+		dsl_dataset_rele(origin, FTAG);
+		dsl_dataset_rele(clone, FTAG);
+	}
+	dsl_pool_rele(dp, FTAG);
+
+	/*
+	 * We don't need to unmount *all* the origin fs's snapshots, but
+	 * it's easier.
+	 */
+	(void) dmu_objset_find(parentname,
+	    zfs_unmount_snap_cb, NULL, DS_FIND_SNAPSHOTS);
+
+	error = dsl_dataset_promote(fsname, conflsnap);
+	if (error != 0)
+		fnvlist_add_string(outnvl, "conflsnap", conflsnap);
+
+	return (error);
+}
+
+static int
+zfs_stable_ioc_set_props(const char *fsname, nvlist_t *innvl, nvlist_t *outnvl,
+    nvlist_t *opts, uint64_t version)
+{
+	zprop_source_t source;
+	boolean_t received;
+	int error = 0;
+
+	received = nvlist_exists(opts, "received");
+	source = (received ? ZPROP_SRC_RECEIVED : ZPROP_SRC_LOCAL);
+	if (received) {
+		nvlist_t *origprops;
+
+		if (dsl_prop_get_received(fsname, &origprops) == 0) {
+			(void) clear_received_props(fsname, origprops, innvl);
+			nvlist_free(origprops);
+		}
+
```

Comment:
I am not sure if we need to support `received` here at all, except maybe for legacy purposes.
Given that `lzc_receive` has an option to set received properties I think that `zc_set_props` should set just local properties.